src/gui/XournalView.cpp

```diff
@@ -86,40 +95,38 @@ void XournalView::staticLayoutPages(GtkWidget* widget, GtkAllocation* allocation
     xv->layoutPages();
 }
 
+
 auto XournalView::clearMemoryTimer(XournalView* widget) -> gboolean {
-    GList* list = nullptr;
+    widget->cleanupBufferCache();
+    return true;
+}
 
-    for (auto&& page: widget->viewPages) {
-        if (page->getLastVisibleTime() > 0) {
-            list = g_list_insert_sorted(list, page, reinterpret_cast<GCompareFunc>(pageViewIncreasingClockTime));
+auto XournalView::cleanupBufferCache() -> void {
+    const auto& [pagesLower, pagesUpper] = this->preloadPageBounds(this->currentPage, this->viewPages.size());
+    g_assert(pagesLower <= pagesUpper);
+
+    // Queue pages to remove, sorted by the oldest (i.e. smallest last visible time)
+    constexpr auto cmp = [](XojPageView* a, XojPageView* b) {
+        return a->getLastVisibleTime() < b->getLastVisibleTime();
+    };
+    std::vector<XojPageView*> toRemove;
+    // Keep in cache: current page + preload_before + preload_after + LRU capacity
+    for (size_t i = 0; i < this->viewPages.size(); i++) {
+        auto&& page = this->viewPages[i];
+        size_t pageNum = i + 1;
+        bool isPreload = pagesLower <= pageNum && pageNum <= pagesUpper;
+        if (page->getLastVisibleTime() > 0 && !isPreload) {
```

Comment:
Flip this for faster comparison. If the first check is already wrong, the second check will not be performed in a `&&`. `!isPreload` is one cycle, compared to a few hundred for accessing the pointer of `page->getLastVisibleTime()`.